src/pint/models/model_builder.py

```diff
@@ -24,295 +25,402 @@
 class UnknownBinaryModel(ValueError):
     """Signal that the par file requested a binary model no in PINT."""
 
+class ComponentConflict(ValueError):
+    """Error for mulitple components can be select but no other indications."""
 
-class ModelBuilder:
-    """A class for model construction interface.
-
-    Parameters
-    ----------
-    name : str
-        Name for the model.
-    parfile : str optional
-        The .par file input for select components. If the parfile is
-        provided the self.model_instance will be put model instance
-        with .par file read in. If it is not provided, self.model_instance
-        will return as None.
 
-    Returns
-    -------
-    A class contains the result model instance if parfile is provided and
-    method to build the model.
-    """
+class ModelBuilder:
+    def __init__(self):
+        self.components = {}
+        for k, v in Component.component_types.items():
+            self.components[k] = v()
+        # The components that always get added.
+        self.default_components = ['SolarSystemShapiro']
+
+    def __call__(self, parfile):
+        param_inpar, repeat_par = self.parse_parfile(parfile)
+        seleted, conflict, unrec_param = self.choose_model(param_inpar)
+        seleted.update(set(self.default_components))
+        # Report conflict
+        if len(conflict) != 0:
+            self._report_conflict(conflict)
+        # Make timing model
+        cps = [self.components[c] for c in seleted]
+        tm = TimingModel(components=cps)
+        # Add indexed parameters
+        leftover_params = set(unrec_param)
+        # Give repeatable parameters an index.
+        for k, v in repeat_par.items():
+            for ii in range(len(v)):
+                leftover_params.add(k + str(ii + 1))
+        tm, unknown_param = self._add_indexed_params(tm, leftover_params)
+        return tm, unknown_param
+
+    @lazyproperty
+    def param_component_map(self):
+        """ Return the parameter to component map.
+        """
+        p2c_map = defaultdict(list)
+        for k, cp in self.components.items():
+            for p in cp.params:
+                p2c_map[p].append(k)
+                # Add alias
+                par = getattr(cp, p)
+                for ap in par.aliases:
+                    p2c_map[ap].append(k)
+        tm = TimingModel()
+        for tp in tm.params:
+            p2c_map[tp].append('timing_model')
+            par = getattr(tm, tp)
+            for ap in par.aliases:
+                p2c_map[ap].append('timing_model')
+        return p2c_map
+
+    @lazyproperty
+    def param_alias_map(self):
+        """ Return the aliases map of all parameters
+        """
+        alias = {}
+        for k, cp in self.components.items():
+            for p in cp.params:
+                par = getattr(cp, p)
+                alias[p] = p
+                for als in par.aliases:
+                    alias[als] = p
```

Comment:
This will not work, since we have components are inherited from the other ones. Some aliases will occur multiple times, but they represent the same parameter. Maybe we can keep track of the value.

Reply:
A new alias that points to a different parameter name is an error; a new occurrence of an alias that points to the same parameter name is fine.